File: sdk/python/aicoder.py
"""AICoder Python SDK

用法:
    from aicoder import AICoder

    client = AICoder(base_url="http://localhost:8787", token="your-token")
    result = client.run("统计 src 下的文件数")
    print(result["text"])
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any, Optional
from urllib import request, error


@dataclass
class RunResult:
    ok: bool
    text: str
    tool_calls: list[str]
    steps: int
    error: Optional[str] = None


class AICoderError(Exception):
    pass


class AICoder:
    """AICoder HTTP 客户端（同步）。"""

    def __init__(
        self,
        base_url: str = "http://localhost:8787",
        token: Optional[str] = None,
        timeout: float = 600.0,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self.token = token
        self.timeout = timeout

    def _headers(self) -> dict[str, str]:
        headers = {"Content-Type": "application/json"}
        if self.token:
            headers["Authorization"] = f"Bearer {self.token}"
        return headers

    def _post(self, path: str, payload: dict[str, Any]) -> dict[str, Any]:
        url = f"{self.base_url}{path}"
        data = json.dumps(payload).encode("utf-8")
        req = request.Request(url, data=data, headers=self._headers(), method="POST")
        try:
            with request.urlopen(req, timeout=self.timeout) as resp:
                return json.loads(resp.read().decode("utf-8"))
        except error.HTTPError as e:
            body = e.read().decode("utf-8", "replace")
            raise AICoderError(f"HTTP {e.code}: {body}") from e
        except error.URLError as e:
            raise AICoderError(f"连接失败: {e.reason}") from e

    def run(self, message: str, use_rag: bool = False, allow_write: bool = False) -> RunResult:
        """一次性执行任务，返回最终文本与统计。"""
        data = self._post(
            "/api/run",
            {"message": message, "useRag": use_rag, "allowWrite": allow_write},
        )
        return RunResult(
            ok=bool(data.get("ok")),
            text=data.get("text", ""),
            tool_calls=data.get("toolCalls", []),
            steps=data.get("steps", 0),
            error=data.get("error"),
        )

    def health(self) -> dict[str, Any]:
        """健康检查。"""
        url = f"{self.base_url}/api/health"
        try:
            with request.urlopen(url, timeout=self.timeout) as resp:
                return json.loads(resp.read().decode("utf-8"))
        except error.URLError as e:
            raise AICoderError(f"连接失败: {e.reason}") from e

File: sdk/go/aicoder.go
// AICoder Go SDK
//
// 用法:
//
//	client := aicoder.NewClient("http://localhost:8787", "your-token")
//	result, err := client.Run(ctx, "统计 src 下的文件数", aicoder.RunOptions{})
//	fmt.Println(result.Text)
package aicoder

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

// Client 是 AICoder HTTP 客户端。
type Client struct {
	BaseURL string
	Token   string
	HTTP    *http.Client
}

// NewClient 创建客户端。
func NewClient(baseURL, token string) *Client {
	return &Client{
		BaseURL: baseURL,
		Token:   token,
		HTTP:    &http.Client{Timeout: 600 * time.Second},
	}
}

// RunOptions 控制单次执行。
type RunOptions struct {
	UseRAG     bool
	AllowWrite bool
}

// RunResult 是一次执行的返回。
type RunResult struct {
	OK        bool     `json:"ok"`
	Text      string   `json:"text"`
	ToolCalls []string `json:"toolCalls"`
	Steps     int      `json:"steps"`
	Error     string   `json:"error,omitempty"`
}

type runRequest struct {
	Message    string `json:"message"`
	UseRAG     bool   `json:"useRag"`
	AllowWrite bool   `json:"allowWrite"`
}

func (c *Client) do(ctx context.Context, method, path string, body interface{}) (*http.Response, error) {
	var buf *bytes.Buffer
	if body != nil {
		data, err := json.Marshal(body)
		if err != nil {
			return nil, err
		}
		buf = bytes.NewBuffer(data)
	} else {
		buf = bytes.NewBuffer(nil)
	}
	req, err := http.NewRequestWithContext(ctx, method, c.BaseURL+path, buf)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	if c.Token != "" {
		req.Header.Set("Authorization", "Bearer "+c.Token)
	}
	return c.HTTP.Do(req)
}

// Run 一次性执行任务。
func (c *Client) Run(ctx context.Context, message string, opts RunOptions) (*RunResult, error) {
	resp, err := c.do(ctx, http.MethodPost, "/api/run", runRequest{
		Message:    message,
		UseRAG:     opts.UseRAG,
		AllowWrite: opts.AllowWrite,
	})
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 300 {
		return nil, fmt.Errorf("HTTP %d", resp.StatusCode)
	}
	var result RunResult
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}
	return &result, nil
}

// Health 健康检查。
func (c *Client) Health(ctx context.Context) (map[string]interface{}, error) {
	resp, err := c.do(ctx, http.MethodGet, "/api/health", nil)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	var out map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
		return nil, err
	}
	return out, nil
}
